openedx/openedx-platform
# refactor: simplify ProblemBlock inheritance hierarchy

## Description

Squashed commit message will be:
```
refactor: simplify ProblemBlock inheritcance hierarchy

Previously, ProblemBlock's implementation was split between
CapaMixin/CapaFields (in capa_base.py) and ProblemBlock
(in capa_module.py), the former being the base classes of the
latter. The split existed for a historical reason:
as a former XModule, ProblemBlock was once split
between CapaDescriptor (author-facing) and CapaModule
(learner-facing). Since ProblemBlock has been converted
to being a pure XBlock, the division between the base
classes and the block class are no longer necessary
nor semantically helpful.

docs: Flesh out ProblemBlock's docstring a bit.
```

Rationale: While exploring the codejail issue, I found that the split between capa_base and capa_module made this code extra difficult to reason about. I had some spare time at the end of the day and took an hour to smash the two modules together.

I broke the change into several sub-commits to make the change easier to review. Despite the big diff, the change is straightforward.

## Testing instructions

Suggested simple smoke test:
* Check out branch.
* `pip install -e common/lib/xmodule` in both lms-shell and studio-shell.
* Restart LMS and Studio.
* Add a problem block to a course, with a couple responses. Publish.
* In LMS, as a student, submit an answer.
* Run an instructor task to rescore the problem. Make sure it succeeds.

## Deadline

None. No urgency.
